### DISCUSSION TITLE
Potential Memory Leak With `build_sensor_for_freshness_checks`

### DISCUSSION THREAD

**Discussion by bolinzzz** [dagster-io/dagster#25276] [Q&A]
After upgrading to 1.8.10, I created a sensor to execute freshness checks based on the instructions [here](https://docs.dagster.io/concepts/assets/asset-checks/checking-for-data-freshness#step-2-create-a-sensor-to-execute-the-check).

After the sensor is deployed, I began observing memory leaks. Screenshot below is showing memory usage of the container running one of our code locations.
![image](https://github.com/user-attachments/assets/04a2561a-b28e-4c75-a9f0-7d643825004a)

This has led me to believe that `build_sensor_for_freshness_checks` is somehow causing the memory leak. 

I haven't done any memory profiling, arrived at this conclusion solely because when I switched off the sensor, the memory leak stopped. 

Some folks were saying the memory leak is caused by `grpcio`, when the grpc client connection fails, see this thread: https://github.com/grpc/grpc/issues/36117

Can someone with more experience help to review the underlying code of `build_sensor_for_freshness_checks`, and identify any potential circumstances that might lead to memory leak. Thank you very much!

**Comment by garethbrickman** [comment]
This may be related to https://github.com/dagster-io/dagster/issues/18997, and https://github.com/dagster-io/dagster/issues/18997#issuecomment-2043485762. What version of grpcio are you using? The solution in there was to pin the grpcio python package to 1.57.0

**Reply by bolinzzz** [reply]
I am using `1.66.2`. Let me pin it to `1.57.0` and will revert back again. Thank you!

**Reply by bolinzzz** [reply]
I tried pinning `grpcio` to `1.57.0` but it did not work. After reducing the evaluation frequency of the sensor from 30s to 20mins helped to stop the leaking.
